File: Chap5/Python/exo4.py
import threading
import random
import time
import sys

class Tab:
    def __init__(self):
        self.tab1 = []
        self.tab2 = []
        self.taille1 = 0
        self.taille2 = 0

def assignation_tab(taille, result, index):
    tableau = []
    for _ in range(taille):
        time.sleep(1)
        tableau.append(random.randint(0, 99))
    result[index] = tableau

def test_inclusion(t):
    compteur = 0
    for i in range(t.taille1):
        for j in range(t.taille2):
            if t.tab1[i] == t.tab2[j]:
                compteur += 1
                break
    if compteur == t.taille1:
        print("le tableau T1 est inclus dans T2")

def wait():
    c = input()
    if c == 'A':
        print("Programme annulé")
        sys.exit(0)


    taille1 = 5
    taille2 = 6

    result = [None, None]
    t = Tab()

    random.seed()

    thread1 = threading.Thread(target=assignation_tab, args=(taille1, result, 0))
    thread2 = threading.Thread(target=assignation_tab, args=(taille2, result, 1))
    thread3 = threading.Thread(target=wait)

    thread1.start()
    thread2.start()
    thread3.start()

    thread1.join()
    thread2.join()

    tab1 = result[0]
    tab2 = result[1]

    t.tab1 = tab1
    t.tab2 = tab2
    t.taille1 = taille1
    t.taille2 = taille2

    thread4 = threading.Thread(target=test_inclusion, args=(t,))
    thread4.start()
    thread4.join()

File: Chap5/C/exo4.c
#include<stdio.h>
#include<pthread.h>
#include<stdlib.h>
#include<time.h>
#include<unistd.h>

struct tab
{
    int *tab1;
    int *tab2;
    int taille1;
    int taille2;
};
typedef struct tab tab;

void *assignation_tab(void*arg)
{ 
    int taille1= *(int*)arg;
    int *tableau=malloc(taille1*sizeof(int));
    for(int i=0;i<taille1;i++)
    {
       sleep(1);
       tableau[i]=rand()%100;
    }
    return (void*)tableau;
}

void *test_inclusion(void *arg)
{
    int compteur=0,i;
    tab t=*(tab*) arg;
    for(i=0;i<t.taille1;i++)
    {
        for(i=0;i<t.taille2;i++)
        {
            if(t.tab1[i]==t.tab2[i])
            {
                compteur++;
            }
        }
    }
    if(compteur==t.taille1)
    {
        printf("le tableau T1 est inclus dans T2");
    }
}

void* wait(void *arg)
{
   int c=getchar();
   if(c=='A')
   {
    printf("Programme annulé");
    exit(0);
   }
   
}

int main()
{
    pthread_t thread1,thread2,thread3,thread4;
    int taille1=5,taille2=6;
    int *tab1,*tab2;
    tab t;
    srand(time(NULL));
    pthread_create(&thread1,NULL,assignation_tab,&taille1);
    pthread_create(&thread2,NULL,assignation_tab,&taille2);
    pthread_create(&thread3,NULL,wait,NULL);
    pthread_join(thread1,(void**)&tab1);
    pthread_join(thread2,(void**)&tab2);
    t.tab1=tab1;
    t.tab2=tab2;
    t.taille1=taille1;
    t.taille2=taille2;
    pthread_create(&thread3,NULL,test_inclusion,&t);
    pthread_join(thread3,NULL);
    
}